consensus/src/aggregation/engine.rs

```diff
@@ -0,0 +1,781 @@
+        let mut locks = Vec::new();
+        let mut acks = Vec::new();
+        let stream = journal
+            .replay(self.journal_replay_concurrency, self.journal_replay_buffer)
+            .await
+            .expect("replay failed");
+        pin_mut!(stream);
+        while let Some(msg_result) = stream.next().await {
+            let (_, _, _, activity) = msg_result.unwrap();
+            match activity {
+                Activity::Tip(index) => {
+                    tip = max(tip, index);
+                }
+                Activity::Lock(item, signature) => {
+                    locks.push((item, signature));
+                }
+                Activity::Ack(ack) => {
+                    acks.push(ack);
+                }
+            }
+        }
+        self.tip = tip;
+        // Add locks
+        locks
+            .iter()
+            .filter(|(item, _)| item.index > tip)
+            .for_each(|(item, signature)| {
+                self.confirmed.insert(item.index, (item.digest, *signature));
+            });
+        // Add any acks that haven't resulted in a threshold signature
+        acks = acks
+            .into_iter()
+            .filter(|ack| ack.item.index > tip && !self.confirmed.contains_key(&ack.item.index))
+            .collect::<Vec<_>>();
+        acks.iter().for_each(|ack| {
+            self.set_rebroadcast_deadline(ack.item.index);
+            assert!(self
+                .pending
+                .insert(
+                    ack.item.index,
+                    Pending::Verified(
+                        ack.item.digest,
+                        HashMap::from([(
+                            ack.epoch,
+                            HashMap::from([(ack.signature.index, ack.clone())]),
+                        )]),
+                    ),
+                )
+                .is_none());
+        });
+    }
+
+    /// Appends an activity to the journal.
+    async fn journal_append(&mut self, activity: Activity<V, D>) {
+        let index = match activity {
+            Activity::Ack(ref ack) => ack.item.index,
+            Activity::Lock(ref item, _) => item.index,
+            Activity::Tip(index) => index,
+        };
+        let section = self.get_journal_section(index);
+        self.journal
+            .as_mut()
+            .expect("journal must be initialized")
+            .append(section, activity)
+            .await
+            .expect("unable to append to journal");
+    }
+
+    /// Syncs (ensures all data is written to disk) and prunes the journal.
+    async fn journal_sync(&mut self, index: Index) {
+        let section = self.get_journal_section(index);
+
+        // Get journal
+        let journal = self.journal.as_mut().expect("journal must be initialized");
+
+        // Sync journal
+        journal.sync(section).await.expect("unable to sync journal");
+
+        // Prune journal, ignoring errors
+        let _ = journal.prune(section).await;
```

Comment:
I think we want to add some buffer for pruning here. If I'm not mistaken, we could get a new ack (near tip + window) that forces us to prune data for something we are still focused on (near tip).

I suggest only calling prune below tip (or even further back).

Reply:
Ah yep, will do below tip

Reply:
https://github.com/commonwarexyz/monorepo/pull/974/commits/e96fb501d1c8dab5edff7ca6e0df00fa40d5e6fb